enhancements/node/lazy-image-pull.md

```diff
@@ -0,0 +1,308 @@
+
+### Drawbacks
+
+#### Resource Consumption with FUSE
+
+The stargz-snapshotter employs a FUSE filesystem, which can lead to increased resource usage on the node during intensive local disk I/O operations. It’s important to note that this does not affect containers performing heavy I/O with local volume mounts, as they will not experience performance degradation. Although in near future this might get mitigated by [FUSE-Passthrough](https://www.phoronix.com/news/Linux-6.9-FUSE-Passthrough)
+
+
+## Test Plan
+
+#### OpenShift End-to-End Testing
+
+Write comprehensive e2e tests to confirm the accurate lazy pulling and execution of containers utilizing `zstd:chunked` images.
+
+#### Performance Benchmarking
+
+Benchmarks comparing the performance of layz pulling `zstd:chunked` images against traditional image usage across diverse operational scenarios.
+
+
+## Graduation Criteria
+
+### Dev Preview -> Tech Preview
+* Introduce support for lazy image pulling per machine config pool in Openshift
+* Chunk verification implemented in `c/storage`
+* Drop-in configuration support in `c/storage`
+* Support lazy pulling from private registry
+* E2e tests in Openshift CI
+
+### Tech Preview -> GA
+* CI test stable for at least one release
+* User facing documentation created in openshift-docs
+
+### Removing a deprecated feature
+
+N/A
+
+## Upgrade / Downgrade Strategy
+
+When an upgrade or downgrade happens with incompatible changes in the store configuration (e.g. enabling/disabling Additional Layer Store), CRI-O handles this by rebuilding its image storage.
+Rebuilding of the image storage is done following the existing `clean_shutdown_file` feature of CRI-O.
+Concretely, CRI-O checks if layers in the storage are damaged and repairs the storage if there are damaged layers.
+
+CRI-O detects incompatibe changes of storage configuration using the following fields of `cri-o.conf`.
+
+```
+layer_integrity_state_file = "<location>"
+layer_integrity_state_marker = "<contents>"
+```
+
+When CRI-O starts for the first time on the node, it creates a file (call it "state file") at the location specified by `layer_integrity_state_file` with the contents of `layer_integrity_state_marker` (call it "state marker").
+Everytime CRI-O restarts, it checks that `layer_integrity_state_marker` in `cri-o.conf` exactly matches the contents of the state file at `layer_integrity_state_file`.
+If it doesn't match, CRI-O treats this as an incompatible change and starts repairing the storage as mentioned above.
+
+MCO populate the state marker with configurations whose changes trigger repairing of the storage, e.g. `"version=1;store-format=1;ALS="estargz:/var/"`.
+The state marker must be an empty string if lazy pulling is disabled and must not be empty if lazy pulling is enabled.
+
+Downgrading CRI-O to a version that doesn't support lazy pulling and state files requires manual wiping of images on the node.
```

Comment:
Cc: @haircommander @saschagrunert 

WDYT about this mechanism? Is it valuable, potentially for other use cases / overengineered / insufficient?

Reply:
yeah this almost seems like something that should happen in a separate service

Reply:
>  (the config file[1] points at the “root” store location

...

> As an analogy, Podman has some kind of “essential config file entries” record in its primary database, and that does include (some of?) the c/storage config settings. I didn’t look into how that works in detail.

Encoding the *file paths* in the storage itself is a broken concept and did break in practice, see https://github.com/containers/podman/pull/23447

But that's distinct from the other config options. I think it could very much make sense to support e.g. `podman image config set composefs=true` which would end up looking up the storage root path from the global config, and then flipping a flag on there that lives in that storage, as opposed to changing the global storage.

In the short term nothing stops crio from doing this itself with `/var/lib/containers/storage/crio-state.json` or whatever.

Reply:
Would that work together with the latest CRI-O wipe enhancements @kwilczynski ?

Reply:
@mtrmac @haircommander 

> a separate service

If this should be a separated service, I think we can have a subcommand in CRI-O, described as the following:

- `crio check-layer-state <layer_integrity_state_marker> <layer_integrity_state_file>`

This will run as a separated service before CRI-O starts.
This does the same things as described in the original proposal. 
Concretely:
- When it starts for the first time on the node, it creates a file (call it "state file") at the location specified by `layer_integrity_state_file` with the contents of `layer_integrity_state_marker` (call it "state marker").
- Everytime this command starts on the node, it checks that the specified `layer_integrity_state_marker` exactly matches the contents of the state file at `layer_integrity_state_file`. If it doesn't match, it starts repairing the storage.

We can also avoid adding `layer_integrity_state_file` and `layer_integrity_state_marker` to crio.conf.

WDYT about this mechanism?

Reply:
that works for me

Reply:
Thanks for the review. Added the description about this subcommand to the proposal.

Reply:
@mtrmac @haircommander I've posted an implementation of this command to https://github.com/cri-o/cri-o/pull/8589 .